File: src/main/java/shopping/product/infrastructure/ProductRepositoryImpl.java
package shopping.product.infrastructure;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Repository;
import shopping.product.domain.Product;
import shopping.product.domain.ProductRepository;

import java.util.Optional;

@Repository
public class ProductRepositoryImpl implements ProductRepository {
    private final ProductJpaRepository repository;

    public ProductRepositoryImpl(ProductJpaRepository repository) {
        this.repository = repository;
    }

    @Override
    public Product save(Product product) {
        return repository.save(ProductEntity.from(product)).toDomain();
    }

    @Override
    public Optional<Product> findById(Long id) {
        return repository.findById(id)
                .map(ProductEntity::toDomain);
    }

    @Override
    public void deleteById(Long id) {
        repository.deleteById(id);
    }

    @Override
    public Page<Product> findAll(Pageable pageable) {
        return repository.findAll(pageable).map(ProductEntity::toDomain);
    }
}

File: src/main/kotlin/shopping/infrastructure/product/ProductRepositoryImpl.kt
package shopping.infrastructure.product

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Component
import shopping.domain.product.Product
import shopping.domain.product.ProductRepository

@Component
class ProductRepositoryImpl(
    private val productJpaRepository: ProductJpaRepository,
) : ProductRepository {
    override fun findAll(pageable: Pageable): Page<Product> = productJpaRepository.findAll(pageable)

    override fun findById(id: Long): Product? = productJpaRepository.findById(id).orElse(null)

    override fun save(product: Product): Product = productJpaRepository.save(product)

    override fun deleteById(id: Long) {
        productJpaRepository.deleteById(id)
    }

    override fun existsById(id: Long): Boolean = productJpaRepository.existsById(id)
}
